kai/service/incident_store/psql.py

```diff
@@ -1,93 +1,182 @@
+class Base(DeclarativeBase):
+    type_annotation_map = {
+        dict[str, Any]: JSON().with_variant(JSONB(), "postgresql"),
+        list[str]: JSON().with_variant(JSONB(), "postgresql"),
+    }
 
-    return wrapper
 
+class PSQLUnmodifiedReport(Base):
+    __tablename__ = "unmodified_reports"
 
-# TODO(@JonahSussman): Migrate this to use an ORM
-class PSQLIncidentStore(IncidentStore):
-    def __init__(self, args: KaiConfigIncidentStorePostgreSQLArgs):
-        self.emb_provider = EmbeddingNone()
+    application_name: Mapped[str] = mapped_column(primary_key=True)
+    generated_at: Mapped[datetime.datetime] = mapped_column(
+        DateTime(), server_default=func.now(), primary_key=True
+    )
+    report: Mapped[dict[str, Any]]
 
-        try:
-            with psycopg2.connect(
-                cursor_factory=DictCursor, **args.model_dump()
-            ) as conn:
-                KAI_LOG.info("Connected to the PostgreSQL server.")
-                self.conn: connection = conn
-                self.conn.autocommit = True
 
-            self.create_tables()
+class ViolationCategory(enum.Enum):
+    potential = "potential"
+    optional = "optional"
+    mandatory = "mandatory"
 
-        except (psycopg2.DatabaseError, Exception) as error:
-            KAI_LOG.error(f"Error initializing PSQLIncidentStore: {error}")
 
-    def create_tables(self):
-        # TODO: Figure out portable way to install the pgvector extension.
-        # Containerize? "CREATE EXTENSION IF NOT EXISTS vector;" Only works as
-        # superuser
+class PSQLApplication(Base):
+    __tablename__ = "applications"
 
-        # TODO: along with analyzer_types.py, we should really use something
-        # like openapi to nail down the spec and autogenerate the types
-        sql_create_tables = open(
-            os.path.join(PATH_SQL, "create_tables.sql"), "r"
-        ).read()
-        sql_add_embedding = open(
-            os.path.join(PATH_SQL, "add_embedding.sql"), "r"
-        ).readlines()
+    application_name: Mapped[str] = mapped_column(primary_key=True)
 
-        with self.conn.cursor() as cur:
-            cur.execute(sql_create_tables)
+    repo_uri_origin: Mapped[str]
+    repo_uri_local: Mapped[str]
+    current_branch: Mapped[str]
+    current_commit: Mapped[str]
+    generated_at: Mapped[datetime.datetime]
 
-            dim = self.emb_provider.get_dimension()
-            for q in sql_add_embedding:
-                cur.execute(q, (dim,))
+    incidents: Mapped[list["PSQLIncident"]] = relationship(
+        back_populates="application", cascade="all, delete-orphan"
+    )
 
-    # Abstract base class implementations
 
-    def delete_store(self):
-        with self.conn.cursor() as cur:
-            cur.execute(open(os.path.join(PATH_SQL, "drop_tables.sql"), "r").read())
+class PSQLRuleset(Base):
+    __tablename__ = "rulesets"
+
+    ruleset_name: Mapped[str] = mapped_column(primary_key=True)
```

Comment:
Rulesets can change over time, not totally sure it matters though

Reply:
oh right since this is just the violations, nvm ignore me. I was thinking if the rules change it would make a difference but nvm